### ISSUE TITLE
ReturnUrl has wrong host name on LoginPageResult on load-balanced environments and external login UI

### ISSUE THREAD

**Issue by snothub** [IdentityServer/IdentityServer4#4038]
**Running on latest version 3.1**

### Issue / Steps to reproduce the problem
TL;DR: when hosting IdSrv in Docker/k8s env with multiple nodes and Interaction service is configured to use a UI hosted on a different domain, /authorize request constructs `ReturnUrl` parameter with wrong host name.

- Calling /authorize for login will result in `return new LoginPageResult(request);`
- LoginPageResult has this code:
```
 if (!loginUrl.IsLocalUrl())
 {
     // this converts the relative redirect path to an absolute one if we're 
     // redirecting to a different server
     returnUrl = context.GetIdentityServerHost().EnsureTrailingSlash() + returnUrl.RemoveLeadingSlash();
 }
```
- `GetIdentityServerHost` extension method uses `request.Host.ToUriComponent()` without taking into consideration that this might just be one of the nodes in the cluster in which case host will just be `localhost`.
```
 public static string GetIdentityServerHost(this HttpContext context)
 {
     var request = context.Request;
     return request.Scheme + "://" + request.Host.ToUriComponent();
 }
```
- I'm thinking `PublicOrigin` should be used here instead.

**Comment by brockallen** [comment]
Do you not have the reverse proxy middleware configured in your IS host? That's the typical way to know the external/real host name.

https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-3.1

**Comment by ilya-chumakov** [comment]
@brockallen  what if a proxy server adds some URL segments in addition to changing the IS hostname?

For example, an external URL for IS could like:

     foo.bar/api/public/auth

so `/authorize` could be accessible as

     foo.bar/api/public/auth/connect/authorize

I assume `X-Forwarded-Host` won't help because it accepts only a hostname (without `/api/public` segment). We fixed the invalid discovery document by setting `PublicOrigin` but still get incorrect `ReturnUrl` in `/authorize` response.

**Comment by brockallen** [comment]
> @brockallen what if a proxy server adds some URL segments in addition to changing the IS hostname?

Then perhaps it should also fix any links rendered in the response payload as well?

**Comment by ilya-chumakov** [comment]
> > @brockallen what if a proxy server adds some URL segments in addition to changing the IS hostname?
> 
> Then perhaps it should also fix any links rendered in the response payload as well?

I'm sure you don't mean it should be done in _real time_ for any response? I'd like to have the links static if possible. Can a custom `ReturnUrl` be configured on IS side?

**Comment by brockallen** [comment]
> I'm sure you don't mean it should be done in real time for any response?

But it's doing it on the request? So either it does it on the response dynamically (and I have seen some products try to do this), or you need something more formal in the HTML generation that knows how to adjust any/all base paths.

**Comment by ilya-chumakov** [comment]
> But it's doing it on the request?

Yes, it is. A proxy _usually_ modifies a limited header subset, leaving a request/response body untouched. However, for a discovery response, we need to read the body, find urls like `foo.bar/connect/token` and replace it with `foo.bar/api/public/connect/token`. So it seems like an _unusual_ task for the proxy (though, no doubt any mature tool can do it). The situation is similar for `/authorize` response. 

I'm just trying to keep all the configuration in one place (IdentityServer) since we've explicitly set `PublicOrigin`. Probably a middleware (modifying `/authorize` response) can do it with no need to re-configure the proxy. Which way would you recommend?

**Comment by snothub** [comment]
Isn't this what the PublicOrigin is for? Could someone explain to me why it
would be wrong to use that and be done with it? Then we wouldn't have to
worry about load balancers at all.

**Comment by leastprivilege** [comment]
PublicOrigin is only for the origin - if paths are being modified by the proxy, IIRC you need to set the right base-path in the IS host.

AFAIK - there are no headers to communicate the base-path from proxy to host. @damianh

**Comment by snothub** [comment]
Right origin is all I need too. Discussion about modifying paths was not
mine and belongs in a separate topic.

**Comment by leastprivilege** [comment]
OK - well - PublicOrigin is a "hack" - but if it works for you, fine.

The "right" way of doing it is using the reverse proxy headers.

**Comment by snothub** [comment]
That's just it - it doesn't work. PublicOrigin is already set, but not used (see first post here). I tried adding:

```
proxy_set_header   X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header   X-Forwarded-Proto $scheme;
```
to the nginx config but that is also not doing any good.

Should I be adding this as well before `.UseIdentityServer()`?
```
app.UseForwardedHeaders(new ForwardedHeadersOptions
{
    ForwardedHeaders = ForwardedHeaders.XForwardedFor | ForwardedHeaders.XForwardedProto
});
```

**Comment by leastprivilege** [comment]
yes. As described in https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-3.1

**Comment by snothub** [comment]
Thanks @leastprivilege . I'm still puzzled by your statement that PublicOrigin is a hack. It feels more like a hack all this I had to go through. And the doc states `If not set, the origin name is inferred from the request.`, so one could argue that this is the one to use if wanting to control the `request.Host`.

But with all the url rewriting of nginx and host being different, I did follow the article. And one more thing I needed to do (because of re-writes) was `app.UsePathBase(new PathString("/idp/root"));`. Now it finally seems to be ok.

**Comment by damianh** [comment]
> 
> 
> PublicOrigin is only for the origin - if paths are being modified by the proxy, IIRC you need to set the right base-path in the IS host.
> 
> AFAIK - there are no headers to communicate the base-path from proxy to host. @damianh

That is correct. This is why I have my an HttpOverrides extension project that supports `X-Forwarded-PathBase` headers https://github.com/ProxyKit/HttpOverrides . You **must** configure your reverse proxy server to forward this header for this to work. 

I don't like using `app.UsePathBase` for reasons described [here](https://github.com/ProxyKit/HttpOverrides#introduction) but I understand that in many situations there is no other choice.
